src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs

```diff
@@ -718,5 +719,105 @@ private async ValueTask<T> CatchAndRethrowExceptionAsync<T>(Func<Task<T>> taskFu
+                  Add to list
+
+            4. Return list
+            */
+            ct.ThrowIfCancellationRequested();
+
+            var singleProjectId = new[] { projectId };
+            var projectsList = await GetProjectsAsync(singleProjectId, ct);
+            var project = projectsList.FirstOrDefault();
+            var packageOrigins = new Dictionary<Tuple<NuGetFramework, string>, IReadOnlyList<IPackageReferenceContextInfo>>();
+
+            if (project != default && project is PackageReferenceProject prProject)
+            {
+                IList<ProjectModel.LockFileTarget> fwGraphList = await prProject.GetFullRestoreGraphAsync(ct);
+
+                if (fwGraphList != null)
+                {
+                    var pkgs = await GetInstalledAndTransitivePackagesAsync(singleProjectId, ct);
+
+                    var visited = new HashSet<object>();
+                    var memory = new Dictionary<object, bool>();
+
+                    foreach (var targetFxGraph in fwGraphList)
+                    {
+                        var key = Tuple.Create(targetFxGraph.TargetFramework, targetFxGraph.RuntimeIdentifier);
+                        var list = new List<IPackageReferenceContextInfo>();
+
+                        foreach (var directPkg in pkgs.InstalledPackages) // are InstalledPackages direct dependencies only? Yes!
+                        {
+                            visited.Clear();
+                            memory.Clear();
+                            var found = FindTransitive(directPkg.Identity, transitivePackage, targetFxGraph, visited, memory);
+                            if (found)
+                            {
+                                list.Add(directPkg);
+                            }
+                        }
+
+                        if (list.Any())
+                        {
+                            packageOrigins[key] = list;
+                        }
+                    }
+                }
+            }
+
+            return packageOrigins;
+        }
+
+        private bool FindTransitive(PackageIdentity current, PackageIdentity transitivePackage, ProjectModel.LockFileTarget graph, HashSet<object> visited, Dictionary<object, bool> memory)
+        {
+            if (current.Equals(transitivePackage))
+            {
+                memory[current] = true;
+                return true;
+            }
+
+            var node = graph
+                .Libraries
+                .Where(x => x.Name == current.Id && x.Version.Equals(current.Version) && x.Type == "package")
+                .FirstOrDefault();
+
+            if (node != default)
+            {
+                visited.Add(current);
+                foreach (var dep in node.Dependencies)
+                {
+                    var pkgChild = new PackageIdentity(dep.Id, dep.VersionRange.MinVersion);
+
+                    if (visited.Contains(pkgChild))
+                    {
+                        return memory[pkgChild];
```

Comment:
`visited` and `memory` could be consolidated. Maybe just using `memory` is enough.

Reply:
If you need 3rd state for value(True/False) is not set yet then maybe `bool?` can be used here. So use `HasValue` check if it's still null or True/False.

Reply:
Thanks! fixed.